### ISSUE TITLE
Postman does not respect the ordering of multi-part form data fields when POSTing

### ISSUE THREAD

**Issue by tomasaschan** [a85/POSTMan-Chrome-Extension#4461]
## Issue Report:
I'm [having trouble with part-ordering of multi-part requests](https://stackoverflow.com/questions/49771811/postman-ordering-of-fields-in-form-data-request), and it seems that Postman _does_ respect the ordering of text parts, but not for file parts. The server I'm using [requires that files are sent last](https://github.com/jaydenseric/graphql-multipart-request-spec#multipart-form-field-structure), but Postman sends files before text parts no matter how I order them in the UI.

## App Details:
```
Postman for Windows
Version 6.0.10
win32 10.0.16299 / x64
```

**Comment by tomasaschan** [comment]
Got some help troubleshooting this with a colleague, who could reproduce the behavior on Mac as well (still version 6.0.10).

With a request body like this:

<img width="1161" alt="postman" src="https://user-images.githubusercontent.com/1550920/38857417-8605bab4-4229-11e8-9254-f8d3490d22b2.png">

the raw HTTP request looks like this:

```
POST / HTTP/1.1
cache-control: no-cache
Postman-Token: d9af53e4-99d8-413d-b05c-ab5f89e152ff
User-Agent: PostmanRuntime/7.1.1
Accept: */*
Host: localhost:9447
accept-encoding: gzip, deflate
content-type: multipart/form-data; boundary=--------------------------150144812641318184457251
content-length: 547
Connection: keep-alive

----------------------------150144812641318184457251
Content-Disposition: form-data; name="0"; filename="test.txt"
Content-Type: text/plain


----------------------------150144812641318184457251
Content-Disposition: form-data; name="operations"

{ "query": "mutation ($file: Upload!) { uploadFile(file: $file) { id } }", "variables": { "file": null } }
----------------------------150144812641318184457251
Content-Disposition: form-data; name="map"

{ "0": ["variables.file"] }
----------------------------150144812641318184457251
```

I would have expected (and needed) the first part to have been last.

**Comment by kamalaknn** [comment]
Hi @tlycken 

I'm trying to reproduce this, but I always see the form body in the order I've specified. Could you share a collection probably using `https://postman-echo.com/post` so that we take a closer look?

**Comment by kamalaknn** [comment]
Hi @tlycken 

I'm closing this due to inactivity. Please provide some resources to reproduce this if this still exists.

**Comment by DHFW** [comment]
@kamalaknn, consider reopening this issue because Postman (6.3.0) definitely has a problem doing this. In Insomnia it works flawlessy.

Curl request (made from form-data in Postman);
```
POST /graphql HTTP/1.1
Host: localhost:x
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW

------WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="operations"

{ "query": "mutation ($file: Upload!) { singleUpload(file: $file) { id } }", "variables": { "file": null } }
------WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="map"

{ "0": ["variables.file"] }
------WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="0"; filename="a.txt"
Content-Type: text/plain


------WebKitFormBoundary7MA4YWxkTrZu0gW--
```

Same way in Insomnia: 

```
> POST /graphql HTTP/1.1
> Host: localhost:x
> User-Agent: insomnia/6.0.2
> Content-Type: multipart/form-data; boundary=X-INSOMNIA-BOUNDARY
> Accept: */*
> Content-Length: 458
| --X-INSOMNIA-BOUNDARY
| Content-Disposition: form-data; name="operations"
| { "query": "mutation ($file: Upload!) { singleUpload(file: $file) { id } }", "variables": { "file": null } }
| --X-INSOMNIA-BOUNDARY
| Content-Disposition: form-data; name="map"
| { "0": ["variables.file"] }
| --X-INSOMNIA-BOUNDARY
| Content-Disposition: form-data; name="0"; filename="a.txt"
| Content-Type: text/plain
| const a = 'Dit is de inhoud van A';
| --X-INSOMNIA-BOUNDARY--
```

Request collection: replace x with a port you use (obfuscated here)

```JSON
{
	"info": {
		"_postman_id": "6df30289-4640-4df9-a94e-b096ec93d55f",
		"name": "Failing request",
		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"
	},
	"item": [
		{
			"name": "http://localhost:x/graphql",
			"request": {
				"method": "POST",
				"header": [],
				"body": {
					"mode": "formdata",
					"formdata": [
						{
							"key": "operations",
							"value": "{ \"query\": \"mutation ($file: Upload!) { singleUpload(file: $file) { id } }\", \"variables\": { \"file\": null } }",
							"type": "text"
						},
						{
							"key": "map",
							"value": "{ \"0\": [\"variables.file\"] }",
							"type": "text"
						},
						{
							"key": "0",
							"value": "",
							"type": "file"
						}
					]
				},
				"url": {
					"raw": "http://localhost:x/graphql",
					"protocol": "http",
					"host": [
						"localhost"
					],
					"port": "x",
					"path": [
						"graphql"
					]
				},
				"description": "Failing post"
			},
			"response": []
		}
	]
}

```

**Comment by kamalaknn** [comment]
Hi @DHFW,

I'm reopening this issue. Can you provide a sample collection or a screencast of this issue? It will help us identify the issue better.

@codenirvana Can you verify this?

**Comment by DHFW** [comment]
@kamalaknn  i have updated the issue with a collection json. It is exactly the same as @tlycken  mentions and shows. However, I'm on Mac latest version.

**Comment by DHFW** [comment]
Another hint might be that when I view the code of the NodeJS request it comes out as the following. It puts the parameter '0' as first, instead of last as in the request editor in Postman (like the screenshot of @tlycken 
```
var fs = require("fs");
var request = require("request");

var options = { method: 'POST',
  url: 'http://localhost:x/graphql',
  headers: { 'content-type': 'multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW' },
  formData: 
   { '0': 
      { value: 'fs.createReadStream("/Users/xxx/back-end/a.txt")',
        options: 
         { filename: '/Users/xxx/back-end/a.txt',
           contentType: null } },
     operations: '{ "query": "mutation ($file: Upload!) { singleUpload(file: $file) { id } }", "variables": { "file": null } }',
     map: '{ "0": ["variables.file"] }' } };

request(options, function (error, response, body) {
  if (error) throw new Error(error);

  console.log(body);
});
```

I was experimenting with this some more with the above code. I placed the `'0'` key as last manually. I console.log-ed the formData.options object and I saw that the options object was sorted again (of course, duh, Javascript object)! Because that's the way an object is sorted... When I changed the `'0'` parameter to `file0` it all started working! I will log an issue in the GraphQL issue for this, to use examples where this could not happen. I don't think Postman can do something about this (if Postman is using Javascript for building the request of course). So consider closing this issue again but at least we have a fix (renaming the parameter of the file)! @tlycken  @kamalaknn

**Comment by DHFW** [comment]
@codenirvana A suggestion for a fix: instead of creating an object for all the formData keys, use 
```
postmanFormData.forEach(key => {
  formData.append(key, value)
})
```

Also try to use this to generate the code for the NodeJS request (because the same problem appears there as well).

**Comment by codenirvana** [comment]
@DHFW Correct, `FormParams` should be stored in `Array` to respect the order and that's how we are storing them in our [collection schema](https://schema.getpostman.com/json/collection/v2.1.0/collection.json).

The issue here is, we use `postman-request`(a fork of [request](https://github.com/request/request)) internally for sending the request which takes `formData` as an object.
We will fix this issue in our fork soon and try to get this fixed in upstream as well.

**Comment by kamalaknn** [comment]
We've updated `postman-runtime` to the version that includes the fix for this in Postman 6.7.1. That should fix the ordering issues you've been seeing.

Please update your app and let us know if you are still seeing this.

**Comment by bhspencer** [comment]
This problem seems to be back in Version 7.34.0 on macOS. I send two parts named "json" and "xyz" and no matter what order I have them in the postman UI the part named "xyz" is sent first.

EDIT: I retract this comment. It turns out that the Java servlet API req.getParts() does not return the parts in order. It is not the fault of postman.
